src/program.cpp

```diff
@@ -501,6 +502,54 @@ std::string perf_group(const operation& op)
     return op.name();
 }
 
+void program::trace(std::ostream& os, parameter_map params) const
+{
+    int n = 1; // collect marker information n times and average it. TODO: do this parameter later
+               // enabling to parse roctx output txt file.
+    uint64_t rangeId = 0; // used as rocTX range number
+    auto& ctx        = this->impl->ctx;
+    // Run once by itself
+    os << "rocTX:\tRunning once..." << std::endl;
+    eval(params);
+    ctx.finish();
+    // dynamically load roctx
+    os << "rocTX:\tLoading rocTX library..." << std::endl;
+    std::filesystem::path fpt           = "/opt/rocm/lib/libroctx64.so";
+    migraphx::dynamic_loader lib_loaded = migraphx::dynamic_loader{fpt};
+    // Instantiate variables for functions.
+    os << "rocTX: linking functions" << std::endl;
+    std::function<void(const char*)> sym_roctxMarkA =
+        lib_loaded.get_function<void(const char*)>("roctxMarkA");
+    std::function<uint64_t(const char*)> sym_roctxRangeStartA =
+        lib_loaded.get_function<uint64_t(const char*)>("roctxRangeStartA");
+    std::function<int(const char*)> sym_roctxRangePushA =
+        lib_loaded.get_function<int(const char*)>("roctxRangePushA");
+    std::function<int()> sym_roctxRangePop = lib_loaded.get_function<int()>("roctxRangePop");
+    std::function<void(uint64_t)> sym_roctxRangeStop =
+        lib_loaded.get_function<void(uint64_t)>("roctxRangeStop");
+    // Fill the map
+    std::unordered_map<instruction_ref, std::vector<double>> ins_vec_roctx;
+    generic_eval(*this, ctx, params, always([&](auto ins, auto) {
+        ins_vec_roctx[ins].reserve(n);
+        return argument{};
+    }));
+    // Run and time each instruction with markers, once
+    sym_roctxMarkA("rocTX: Marker");
+    rangeId = sym_roctxRangeStartA("rocTX: RangeStart");
+    generic_eval(*this, ctx, params, always([&](auto ins, auto f) {
+        argument result;
+        std::cout << "rocTX:\tMarker start:\t" << ins->name() << std::endl;
+        ins_vec_roctx[ins].push_back(time<milliseconds>([&] {
```

Comment:
The `push_back` and `time` function can be removed, as its not needed since we are using roctracer to track the times.

Reply:
I have actually intentionally left this, I was curious if I have let this time function here and call tracers/markers inside, then if I executed perf_report where it just does time() function, the difference between the two should give me the overhead of calling these markers.

Reply:
Though I agree it is not needed in released code, as this is more of a debugging/understanding study code.